### ISSUE TITLE
Add ability to create gin indexes on specific jsonb fields

### ISSUE THREAD

**Issue by barryhagan** [JasperFx/marten#1190]
For an object in my database, I need to use a containment query like this:
```
var matches = new[] {"abc", "xyz"};
context.Query<T>().Where(x => x.MemberIds.Any(m => matches.Contains(m)));
```
Marten has specific support for this query, and produces a Postgres search like this:
```
data -> 'MemberIds' ?| array['abc', 'xyz']`
```

The problem is that this type of query will not be able to use either a ComputedIndex on this field via `For<T>.Index(x => x.MemberIds)`, nor the complete gin index on data that can be added using `For<T>.GinIndexJsonData()`.

The computed index does not work because it selects using `data ->> MemberIds` and the gin index does not work because the query needs `data -> MemberIds` instead of just `data`.

It seems like the best solution here would be the ability to create an `IndexDefinition` instead of a `ComputedIndex` for this field, which could be setup as gin with jsonb_ops to support the query.

Right now the only way I've found to hack this is to directly modify the Expression on the gin index of data, so the specific field is also available with jsonb_ops instead of jsonb_path_ops to support the query:

```
// ugly, ugly hack
For<T>().GinIndexJsonData(def => 
 def.Expression = "data jsonb_path_ops, ((data -> 'MemberIds'::text))"
);
```
This hack is very fragile because the formatting of the Expression is used to when doing a SchemaUpdate check.  Those extra parens around the field and spacing is critical or it will try to rebuild the index on each schema check.

Ideally I'd like to have something like this which would create an index that supports `?|` for the specific Any().Contains() query on MemberIds
```
public class DocumentMappingExpression<T>
{
    public DocumentMappingExpression<T> GinIndex(Expression<Func<T, object>> expression, Action<IndexDefinition> configure = null);
}
```

**Comment by oskardudycz** [comment]
@barryhagan  I think that what you're trying to achieve is already supported. See `Customizing a Calculated Index` section of documentation: http://jasperfx.github.io/marten/documentation/documents/customizing/computed_index/

So you could do:

```
var store = DocumentStore.For(_ =>
{
    _.Connection(ConnectionSource.ConnectionString);

    // The second, optional argument to Index()
    // allows you to customize the calculated index
    _.Schema.For<YourTypes>().Index(x => x.MemberIds, x =>
    {
        // Change the index method to "gin"
        x.Method = IndexMethod.gin;
    });
});

```

**Comment by barryhagan** [comment]
@oskardudycz  that won't work because of how the locator is built between index and query time.

If I setup the index like you suggest, it will create an index with this DDL:
```
CREATE INDEX mt_doc_type_idx_member_ids ON public.mt_doc_type USING gin (  (CAST(data ->> 'MemberIds' as jsonb)));
```

But the query is always performed using:
```
data -> 'MemberIds' ?| array['abc', 'xyz']`
```

Those will not match each other and so it drops to a sequential scan.

I need the index to be created with the same locator syntax as the query, or alternatively it would work if the query was changed to use the same CAST locator as the index:
```
where (CAST(data->>'MemberIds' as jsonb) ?| array['xyz'] )
```

So maybe the actual request here is to change the .Any().Contains() query to match the index definition. But I'm not sure why the selection as string and cast back to jsonb is required.

**Comment by mysticmind** [comment]
Closed via #1334
